Source/WTF/wtf/Vector.h

```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
-    }
-
-    static void initializeIfNonPOD(T* begin, T* end)
-    {
-        VectorInitializer<VectorTraits<T>::needsInitialization, VectorTraits<T>::canInitializeWithMemset, T>::initializeIfNonPOD(begin, end);
-    }
-
-    static void initialize(T* begin, T* end)
-    {
-        VectorInitializer<VectorTraits<T>::needsInitialization, VectorTraits<T>::canInitializeWithMemset, T>::initialize(begin, end);
-    }
-
-    template<typename ... Args>
-    static void initializeWithArgs(T* begin, T* end, Args&&... args)
-    {
-        VectorInitializer<VectorTraits<T>::needsInitialization, VectorTraits<T>::canInitializeWithMemset, T>::initializeWithArgs(begin, end, std::forward<Args>(args)...);
-    }
-
-    static void move(T* src, T* srcEnd, T* dst)
-    {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::move(src, srcEnd, dst);
-    }
 
     static void move(std::span<T> src, std::span<T> dst)
     {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::move(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
+        move(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
     }
 
     static void moveOverlapping(T* src, T* srcEnd, T* dst)
     {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::moveOverlapping(src, srcEnd, dst);
+        if constexpr (VectorTraits<T>::canMoveWithMemcpy)
+            memmove(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
+        else {
+            if (src > dst)
+                move(src, srcEnd, dst);
+            else {
+                T* dstEnd = dst + (srcEnd - src);
+                while (src != srcEnd) {
+                    --srcEnd;
+                    --dstEnd;
+                    new (NotNull, dstEnd) T(WTFMove(*srcEnd));
+                    srcEnd->~T();
+                }
+            }
+        }
     }
 
     static void moveOverlapping(std::span<T> src, std::span<T> dst)
     {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::moveOverlapping(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
+        moveOverlapping(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
     }
 
     static void uninitializedCopy(const T* src, const T* srcEnd, T* dst)
     {
-        VectorCopier<VectorTraits<T>::canCopyWithMemcpy, T>::uninitializedCopy(src, srcEnd, dst);
+        if constexpr (VectorTraits<T>::canCopyWithMemcpy)
+            memcpy(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
+        else {
+            while (src != srcEnd) {
+                new (NotNull, dst) T(*src);
+                ++dst;
+                ++src;
+            }
```

Comment:
I think we can use `std::uninitialized_copy` or `std::ranges_uninitialized_copy` here instead of writing our own loop, although we will need to verify that it's still equally efficient.

Reply:
This code is old which has a lot of problems. I'm focusing on TMP in this change. out of scope of this change.